python/ray/tune/suggest/search_generator.py

```diff
@@ -0,0 +1,209 @@
+            trial = self.create_trial_if_possible(self._experiment.spec,
+                                                  self._experiment.name)
+            if trial is None:
+                break
+            trials.append(trial)
+        return trials
+
+    def create_trial_if_possible(self, experiment_spec, output_path):
+        logger.debug("creating trial")
+        trial_id = Trial.generate_id()
+        suggested_config = self.searcher.suggest(trial_id)
+        if suggested_config == Searcher.FINISHED:
+            self._finished = True
+            logger.debug("Searcher has finished.")
+            return
+
+        if suggested_config is None:
+            return
+        spec = copy.deepcopy(experiment_spec)
+        spec["config"] = merge_dicts(spec["config"],
+                                     copy.deepcopy(suggested_config))
+
+        # Create a new trial_id if duplicate trial is created
+        flattened_config = resolve_nested_dict(spec["config"])
+        self._counter += 1
+        tag = "{0}_{1}".format(
+            str(self._counter), format_vars(flattened_config))
+        trial = create_trial_from_spec(
+            spec,
+            output_path,
+            self._parser,
+            evaluated_params=flatten_dict(suggested_config),
+            experiment_tag=tag,
+            trial_id=trial_id)
+        return trial
+
+    def on_trial_result(self, trial_id, result):
+        """Notifies the underlying searcher."""
+        self.searcher.on_trial_result(trial_id, result)
+
+    def on_trial_complete(self, trial_id, result=None, error=False):
+        self.searcher.on_trial_complete(
+            trial_id=trial_id, result=result, error=error)
+
+    def is_finished(self):
+        return self._counter >= self._total_samples or self._finished
+
+    def get_state(self):
+        return {
+            "counter": self._counter,
+            "total_samples": self._total_samples,
+            "finished": self._finished,
+            "experiment": self._experiment
+        }
+
+    def set_state(self, state):
+        self._counter = state["counter"]
+        self._total_samples = state["total_samples"]
+        self._finished = state["finished"]
+        self._experiment = state["experiment"]
+
+    def has_checkpoint(self, dirpath):
+        return bool(
+            _find_newest_ckpt(dirpath, self.CKPT_FILE_TMPL.format("*")))
+
+    def save_to_dir(self, dirpath, session_str):
+        searcher = self.searcher
+        search_alg_state = self.get_state()
+        while hasattr(searcher, "searcher"):
+            searcher_name = type(searcher).__name__
+            if searcher_name in search_alg_state:
+                logger.warning(
+                    "There was a duplicate when saving {}. "
+                    "Restore may not work properly.".format(searcher_name))
+            else:
+                search_alg_state["name:" +
+                                 searcher_name] = searcher.get_state()
```

Comment:
Should we add a `get_state` with `raise NotImplementedError` to the base searcher class? I get that `get_state` won't get called if there is no `searcher` attribute in the class, but since we use the abstract searcher interface it might be good to define all methods there anyway.